pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,276 @@
+	errEmptyFilename           = errors.New("empty filename")
+	errEmptyPath               = errors.New("empty path")
+	errEmptyDirectory          = errors.New("empty directory")
+	errInvalidArg              = errors.New("invalid filename/directory")
+	directoryAlreadyExistError = "550 Create directory operation failed."
+	transferComplete           = "226 Transfer complete."
+)
+
+// ftpFileSystem represents a file system interface over FTP.
+type ftpFileSystem struct {
+	*ftpFile
+	conn   ServerConn
+	config *Config
+	logger Logger
+}
+
+// Config represents the FTP configuration.
+type Config struct {
+	Host      string // FTP server hostname
+	User      string // FTP username
+	Password  string // FTP password
+	Port      int    // FTP port
+	RemoteDir string // Remote directory path. Base Path for all FTP Operations.
+}
+
+// New initializes a new instance of ftpFileSystem with provided configuration.
+func New(config *Config) FileSystem {
+	return &ftpFileSystem{config: config}
+}
+
+// UseLogger sets the logger interface for the FTP file system.
+func (f *ftpFileSystem) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		f.logger = l
+	}
+}
+
+// UseMetrics sets the metrics for the ftpFileSystem client which asserts the Metrics interface.
+// Currently not implemented.
+func (*ftpFileSystem) UseMetrics(_ interface{}) {
+
+}
+
+// Connect establishes a connection to the FTP server and logs in.
+func (f *ftpFileSystem) Connect() {
+	ftpServer := fmt.Sprintf("%v:%v", f.config.Host, f.config.Port)
+
+	const dialTimeout = 5 * time.Second
+
+	conn, err := ftp.Dial(ftpServer, ftp.DialWithTimeout(dialTimeout))
+	if err != nil {
+		f.logger.Errorf("Failed to connect to FTP server: %v", err)
+		return
+	}
+
+	f.conn = &Conn{conn}
+
+	err = conn.Login(f.config.User, f.config.Password)
+	if err != nil {
+		f.logger.Errorf("Failed to login: %v", err)
+	}
+
+	f.logger.Logf("Login Successful")
+}
+
+// Create creates an empty file on the FTP server.
+func (f *ftpFileSystem) Create(name string) (File, error) {
+	if name == "" {
+		return nil, errEmptyFilename
+	}
+
+	emptyReader := new(bytes.Buffer)
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.Stor(filePath, emptyReader)
+	if err != nil {
+		return nil, err
+	}
+
+	f.logger.Logf("Created %s", name)
+
+	s := strings.Split(filePath, "/")
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s[len(s)-1],
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+	}, nil
+}
+
+// Mkdir creates a directory on the FTP server. Here, os.FileMode is unused.
+func (f *ftpFileSystem) Mkdir(name string, _ os.FileMode) error {
+	if name == "" {
+		return errEmptyDirectory
+	}
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
+
+	err := f.conn.MakeDir(filePath)
+	if err == nil {
+		f.logger.Logf("Created %s directory", name)
+		return nil
+	}
```

Comment:
[unusual] Normally one does `if err != nil { … }` to keep the non-error path unindented.